Improved ui org tests

renzon opened the pull request:
close #2971

test_positive_remove* tests reusing same org

deleted test_positive_add* already tested on test_positive_remove*:

test_positive_add_domain
test_positive_add_user
test_positive_add_hostgroup
test_positive_add_compresource
test_positive_add_medium
test_positive_add_template
test_positive_add_environment
test_positive_add_subnet

Test results before PR:


```console
(robottelo) [renzo@note robottelo]$ pytest tests/foreman/ui/test_organization.py -k "OrganizationTestCase and test_positive_remove"
============================= test session starts ==============================
platform linux2 -- Python 2.7.12, pytest-3.0.7, py-1.4.32, pluggy-0.4.0
rootdir: /home/renzo/PycharmProjects/robottelo, inifile:
plugins: xdist-1.15.0, services-1.1.14, cov-2.4.0
collected 33 items 
2017-03-16 12:00:32 - conftest - DEBUG - Found WONTFIX in decorated tests ['1269196', '1245334', '1217635', '1226425', '1156555', '1204686', '1267224', '1103157', '1230902', '1214312', '1079482']

2017-03-16 12:00:32 - conftest - DEBUG - Collected 33 test cases

2017-03-16 12:00:32 - conftest - DEBUG - Deselected test tests.foreman.ui.test_organization.test_positive_create_with_auto_populated_label due to WONTFIX
tests/foreman/ui/test_organization.py ......s...

============================= 23 tests deselected ==============================
============ 9 passed, 1 skipped, 23 deselected in 1180.65 seconds =============
```
After PR:
```console
(robottelo) [renzo@note robottelo]$ pytest tests/foreman/ui/test_organization.py -k "OrganizationTestCase and test_positive_remove"
============================= test session starts ==============================
platform linux2 -- Python 2.7.12, pytest-3.0.7, py-1.4.32, pluggy-0.4.0
rootdir: /home/renzo/PycharmProjects/robottelo, inifile:
plugins: xdist-1.15.0, services-1.1.14, cov-2.4.0
collected 25 items 
2017-03-16 13:12:03 - conftest - DEBUG - Found WONTFIX in decorated tests ['1269196', '1245334', '1217635', '1226425', '1156555', '1204686', '1267224', '1103157', '1230902', '1214312', '1079482']

2017-03-16 13:12:03 - conftest - DEBUG - Collected 25 test cases

2017-03-16 13:12:03 - conftest - DEBUG - Deselected test tests.foreman.ui.test_organization.test_positive_create_with_auto_populated_label due to WONTFIX


tests/foreman/ui/test_organization.py ......s...

============================= 15 tests deselected ==============================
============ 9 passed, 1 skipped, 15 deselected in 1031.78 seconds =============
(robottelo) [renzo@note robottelo]$ 

```

So execution time is improved ~ 13%, not counting the time of removed tests ;)

rochacbruno commented:
The same problem has been worked here: https://github.com/SatelliteQE/robottelo/issues/3519

We have the `set_session_org` approach in place, if moving we need to stick with a single approach, either removing the existing `set_session_org` or moving your implementation to that patter.
 https://github.com/SatelliteQE/robottelo/blob/master/robottelo/test.py#L223-L248

rochacbruno commented on tests/foreman/ui/test_organization.py:
```diff
@@ -365,13 +382,15 @@ def test_positive_remove_domain(self):
         """
+
         with Session(self.browser) as session:
+            org_name = _org_name_module_scope(session)
```
I like your implementation as it is clearer than current `set_session_org`, my suggestion s to bind the cache getter to the session.

instead of:

```python
org_name = _org_name_module_scope(session)
```

we do:

```python
org = session.get_org() 
org_name = session.get_org_name() 
```

and this uses the same `_org_name_module_scope` cache factory.

renzon replied:
I made the modification and will post the result. It's more clear and allow org reuse on all other UI tests as well. ty.

renzon replied:
Modified tests passed:

```console
(robottelo) [renzo@note robottelo]$ pytest tests/foreman/ui/test_organization.py -k "OrganizationTestCase and test_positive_remove"
============================= test session starts ==============================
platform linux2 -- Python 2.7.12, pytest-3.0.7, py-1.4.32, pluggy-0.4.0
rootdir: /home/renzo/PycharmProjects/robottelo, inifile:
plugins: xdist-1.15.0, services-1.1.14, cov-2.4.0
collected 25 items 
2017-03-16 14:20:11 - conftest - DEBUG - Found WONTFIX in decorated tests ['1269196', '1245334', '1217635', '1226425', '1156555', '1204686', '1267224', '1103157', '1230902', '1214312', '1079482']

2017-03-16 14:20:11 - conftest - DEBUG - Collected 25 test cases

2017-03-16 14:20:11 - conftest - DEBUG - Deselected test tests.foreman.ui.test_organization.test_positive_create_with_auto_populated_label due to WONTFIX


tests/foreman/ui/test_organization.py ......s...

============================= 15 tests deselected ==============================
============ 9 passed, 1 skipped, 15 deselected in 1130.09 seconds =============
```

oshtaier commented:
I would say that implemented logic is a point for discussion as you completely changing idea for test cases. From test automation perspective, it is incorrect to substitute create procedure on update as you start mixing two independent functionalities under one test case. So if removal procedure contains a defect, you will mask success results for addition functionality. Test cases granulation treated as more important criteria than any performance improvement.

From my perspective, it is NACK, but of course, it will be nice to hear any other opinions

renzon commented:
@oshtaier to clarify, so you have problem with the add tests I have deleted, right? But not reusing same org on all remove* tests?

On granularity of test, let's use an example. remove_domain, add_domain. To test remove, you need to be able to add and that is why "add" code was duplicated on "remove". In this case if "add" fails we would have 2 failures, once both tests depend on add feature. If remove fail, just one test failing and the "add" test would be green. On the other hand if we have only remove, only on test will fail or pass. Besides it can be an add problem here, the fail will have the line which is failing what will be quickly checked by QE and BZ can be filled accordingly. The only good side of running this duplicated code to me is improving PQI. So it would be a "practicality beats purity". But I can revert if other folks share your opinion ;)

oshtaier commented:
@renzon 

1):
```
-    make_org(session, org_name=org_name, users=[user_name])
```
and
```
 +    self.org.update(
 +        org_name, users=[user_name], new_users=None,
 +        select=True
 +    )
```
are completely different things from test perspective as in first case you add user on create phase, and in second case - on update phase.

2) We have to have next scenarios:
a) Create organization and add new user to it. Verify that it is added
b) Create organization with user and try to remove user from that organization
c) (Optional and redundant) Create organization with new user
d) (Optional and redundant) Create organization, add new user and remove new user

We cannot mix a) and b), because it is just wrong from QA perspective

renzon commented:
@oshtaier got what you mean. So I checked all add* tests and changed them to create_with*. I have also made sure those were not using org.update.

I also renamed all remove* to update* tests and made sure they all only use org.update to add and remove properties. For those I reused same org all the time.

Tests results:

```console
============================= test session starts ==============================
platform linux2 -- Python 2.7.12, pytest-3.0.7, py-1.4.32, pluggy-0.4.0
rootdir: /home/renzo/PycharmProjects/robottelo, inifile:
plugins: xdist-1.15.0, services-1.1.14, cov-2.4.0
collected 35 items 
2017-03-20 13:53:28 - conftest - DEBUG - Found WONTFIX in decorated tests ['1269196', '1378009', '1245334', '1221971', '1217635', '1226425', '1156555', '1204686', '1267224', '1103157', '1230902', '1214312', '1079482']

2017-03-20 13:53:28 - conftest - DEBUG - Collected 35 test cases

2017-03-20 13:53:28 - conftest - DEBUG - Deselected test tests.foreman.ui.test_organization.test_positive_create_with_auto_populated_label due to WONTFIX


tests/foreman/ui/test_organization.py .............s..........F...s...FF

=================================== FAILURES ===================================
______________ OrganizationTestCase.test_positive_update_location ______________
______________ OrganizationTestCase.test_verify_bugzilla_1225588 _______________
______________ OrganizationTestCase.test_verify_bugzilla_1259248 _______________
```

From 3 failures, I have touched only the first one, update location. Running it again:

```console
home/renzo/PycharmProjects/venvs/robottelo/bin/python /home/renzo/bin/pycharm-2016.3/helpers/pycharm/pytestrunner.py -p pytest_teamcity /home/renzo/PycharmProjects/robottelo/tests/foreman/ui/test_organization.py "-k OrganizationTestCase and test_positive_update_location"
Testing started at 2:55 PM ...
2017-03-20 14:55:08 - conftest - DEBUG - Registering custom pytest_namespace

============================= test session starts ==============================
============================= 34 tests deselected ==============================
================== 1 passed, 34 deselected in 159.64 seconds ===================

Process finished with exit code 0
```

renzon commented:
@rochacbruno can you review it again once I made lot of changes based on @oshtaier's comments?

rochacbruno commented on tests/foreman/ui/test_organization.py:
```diff
@@ -355,23 +355,121 @@ def test_negative_update(self):
+                    org_name = gen_string('alpha')
+                    make_org(session, org_name=org_name, subnets=[subnet_name])
+                    self.assertIsNotNone(self.org.search(org_name))
+                    # self.org.update(org_name, new_subnets=[subnet_name])
```
This update should be commented out?

rochacbruno commented on tests/foreman/ui/test_organization.py:
```diff
@@ -355,23 +355,121 @@ def test_negative_update(self):
+                    # Item is listed in 'Selected Items' list and not
+                    # 'All Items' list.
+                    self.assertIsNotNone(element)
+                    self.org.update(org_name, subnets=[subnet_name])
```
I understood that where we `update` the org we should always create a new one and not reuse from cache.

renzon replied:
@rochacbruno. I followed the code and actually saw. On "update" you go to Manage orgs and search for "org_name" on existing orgs list. On the other hand "make_org" really create a new one. That is why using the two of them is different test cases as @oshtaier pointed out.

renzon replied:
@rochacbruno On update tests always add and remove entities related to Org. And once each test only check the property under test, one failing test will not impact another. But I can be sure for the tests unders this module, it could maybe impact others that start to use it. So, like you, I am going to wait for @oshtaier's opinion ;)

oshtaier replied:
Of course, implemented logic is much better than your previous version, but still it is arguable point whether performance improvement worth losing proper test case isolation or not

rochacbruno commented:
Hi @renzon from my side it is ACK. Now waiting for @oshtaier review.

oshtaier commented on robottelo/ui/session.py:
```diff
@@ -46,3 +50,19 @@ def logout(self):
+    def get_org_name(self):
+        """
+        Make a org and cache it to be used on test session, avoiding
+        overhead of org creation on each test. At the end of test org
```
`At the end of test, organization must be at same state (not mutated)`

oshtaier commented on robottelo/ui/session.py:
```diff
@@ -46,3 +50,19 @@ def logout(self):
+        overhead of org creation on each test. At the end of test org
+        Must be at same state, e.g., you must not mutate org's state.
+
+        If mutation is need create your own on test itself.
```
`If mutation is needed -  create your own organization on test itself.`

oshtaier commented on robottelo/ui/session.py:
```diff
@@ -46,3 +50,19 @@ def logout(self):
+
+        If mutation is need create your own on test itself.
+
+        :return: str with Organization's name
```
`:return str: Organization name`

oshtaier commented on tests/foreman/ui/test_organization.py:
```diff
@@ -355,23 +355,120 @@ def test_negative_update(self):
+                    make_org(session, org_name=org_name, subnets=[subnet_name])
+                    self.assertIsNotNone(self.org.search(org_name))
+                    self.org.search_and_click(org_name)
+                    session.nav.click(tab_locators['context.tab_subnets'])
```
Suggest to use `self.org` instead `session.nav`, but both option gonna work fine

renzon commented:
@oshtaier thinking about what you said I am in doubt as well. In my first approach I used a module level cache organization. This provided more confidence in isolation once I am sure update tests under it will not affect each other even when one of them fail to remove some property. Do you think I should revert it back?

renzon commented:
@oshtaier put what said into code: https://github.com/renzon/robottelo/commit/31f8c3caf260fcaa8e2d7bb1af9b3e2b5be92b3c.

renzon commented:
@abalakh: @rochacbruno mentionad same thing. But problem is that in this case I am changing org properties and messing with [default org](https://github.com/renzon/robottelo/blob/cb61df0ccb1ec8fea0c597e6c50c487b79e81bf6/robottelo/test.py#L228-L228) would also broke test isolation mentioned by @oshtaier.

renzon commented:
@oshtaier @abalakh I reverted org creation to each update* test case. Can you guys review?

ldjebran commented on robottelo/ui/session.py:
```diff
@@ -46,3 +50,21 @@ def logout(self):
+        if 'org_name' in _org_cache:
+            return _org_cache['org_name']
+        _org_cache['org_name'] = gen_string('alpha')
+        make_org(self, org_name=_org_cache['org_name'])
```
put the org in cache after it gets really created eg: inversed
```python
org_name = gen_string('alpha')
make_org(self, org_name=org_name)
_org_cache['org_name'] = org_name
```

in case of a failure in make_org, we still have a chance to create it, but this way, the org name will be in the cache but the org may not exist

ldjebran replied:
```python
org_name = _org_cache.get('org_name')
if org_name is not None:
    return org_name
org_name = gen_string('alpha')
make_org(self, org_name=org_name)
_org_cache['org_name'] = org_name
return org_name


```

oshtaier commented:
Okay, my opinion:
1) Leave cache method as you implemented it
2) Leave sorting for tests (add, delete, add, delete...)
3) Leave all docstring updates
4) Leave any fixes in test logic
5) All other code changes will be fine to revert to its previous state (15 or 25% of performance improve doesn't worth proper test isolation)

abalakh commented on tests/foreman/ui/test_organization.py:
```diff
@@ -376,41 +432,133 @@ def test_negative_update(self):
+                    self.assert_org_update_with(
+                        org_name, session, subnet, subnet_name,
+                        'context.tab_subnets', subnets=[subnet_name]
+                    )
```
Personally i'm not too happy about such assertions. We're always trying to keep tests as much simple as possible and such custom assertions aren't clear at all - without looking through their implementation i can't say what's going on in this test.
Just take a look - it's called `assert_org_update_with` but according to docstring test is not updating but removing the subnet, those args are not user-friendly too - helper accepts everything starting from session and ending with string containing key to locator of some tab.
I've recently had a discussion with @rplevka and his point was to avoid introducing such magic helpers to keep test steps clear without need to analyze helpers source code to understand them. Summoning @rplevka here too. :) Roman, what do you think about this case?

renzon replied:
@abalakh I totally agree on naming been more explicitly and most of time I struggle creating it. 

But the logic on all tests was very clear: Create Org, updating adding related entitiy, update removing entity. So I can't see how extracting this logic can be more difficult than CTLR+C  CTLR+V the entire logic over tests, so against DRY. Once you read helper assert for the first time, on first test, you can understand it for all other tests. The upside of this approach is, more than just having less code, update on assertion can be done in just one place. Just like know that we decided not to cache org on updates, I just changed that on helper. The first time I made the change to use cache I had to go over all tests and changing each on on the same way, risking forget some of them. In fact, this is exactly the problem with breaking

So I really like @SatelliteQE/quality-engineers opinion. I can revert it in case most think copy and paste is better, even without not at all agreeing to it ;)

abalakh replied:
Sure, as i said, that was just my imo. Let's hear other folks.

oshtaier commented:
Here is overall idea that I have in mind which is basically in our code style and can remove redundancy in code a lot:
https://gist.github.com/oshtaier/8a181b5f379565b29fd41905dfc21218

renzon commented:
@oshtaier very nice approach, I am going to try it for this scenario.

oshtaier commented:
I understand the reason to put so-called helpers into test module itself as in that case you remove even more redundant code, but if you put it on a higher level (like in 'org.py') then that code can be reused in any module we like to add or remove something from organization.

renzon commented:
@oshtaier I refactored the code to org and it really got simplified once locators are available there. After that I undestood what @abalakh mean with factory methods, going to keep it in my mind for future enhancements. I will post test results here once they are done.

renzon commented:
Tests Results

```console
pytest tests/foreman/ui/test_organization.py -v -k "OrganizationTestCase"

tests/foreman/ui/test_organization.py::OrganizationTestCase::test_negative_create PASSED
tests/foreman/ui/test_organization.py::OrganizationTestCase::test_negative_create_with_same_name FAILED
tests/foreman/ui/test_organization.py::OrganizationTestCase::test_negative_update PASSED
tests/foreman/ui/test_organization.py::OrganizationTestCase::test_positive_create_with_both_loc_and_org PASSED
tests/foreman/ui/test_organization.py::OrganizationTestCase::test_positive_create_with_compresource <- robottelo/decorators/__init__.py PASSED
tests/foreman/ui/test_organization.py::OrganizationTestCase::test_positive_create_with_domain <- robottelo/decorators/__init__.py PASSED
tests/foreman/ui/test_organization.py::OrganizationTestCase::test_positive_create_with_environment <- robottelo/decorators/__init__.py PASSED
tests/foreman/ui/test_organization.py::OrganizationTestCase::test_positive_create_with_hostgroup <- robottelo/decorators/__init__.py PASSED
tests/foreman/ui/test_organization.py::OrganizationTestCase::test_positive_create_with_location <- robottelo/decorators/__init__.py FAILED
tests/foreman/ui/test_organization.py::OrganizationTestCase::test_positive_create_with_media <- robottelo/decorators/__init__.py PASSED
tests/foreman/ui/test_organization.py::OrganizationTestCase::test_positive_create_with_name PASSED
tests/foreman/ui/test_organization.py::OrganizationTestCase::test_positive_create_with_ptable <- robottelo/decorators/__init__.py PASSED
tests/foreman/ui/test_organization.py::OrganizationTestCase::test_positive_create_with_same_name_and_label PASSED
tests/foreman/ui/test_organization.py::OrganizationTestCase::test_positive_create_with_smartproxy <- robottelo/decorators/__init__.py SKIPPED
tests/foreman/ui/test_organization.py::OrganizationTestCase::test_positive_create_with_template <- robottelo/decorators/__init__.py PASSED
tests/foreman/ui/test_organization.py::OrganizationTestCase::test_positive_create_with_unmatched_name_label PASSED
tests/foreman/ui/test_organization.py::OrganizationTestCase::test_positive_create_with_user PASSED
tests/foreman/ui/test_organization.py::OrganizationTestCase::test_positive_delete PASSED
tests/foreman/ui/test_organization.py::OrganizationTestCase::test_positive_search_autocomplete PASSED
tests/foreman/ui/test_organization.py::OrganizationTestCase::test_positive_update_compresource <- robottelo/decorators/__init__.py PASSED
tests/foreman/ui/test_organization.py::OrganizationTestCase::test_positive_update_domain <- robottelo/decorators/__init__.py PASSED
tests/foreman/ui/test_organization.py::OrganizationTestCase::test_positive_update_environment <- robottelo/decorators/__init__.py PASSED
tests/foreman/ui/test_organization.py::OrganizationTestCase::test_positive_update_hostgroup <- robottelo/decorators/__init__.py FAILED
tests/foreman/ui/test_organization.py::OrganizationTestCase::test_positive_update_location <- robottelo/decorators/__init__.py PASSED
tests/foreman/ui/test_organization.py::OrganizationTestCase::test_positive_update_media <- robottelo/decorators/__init__.py PASSED
tests/foreman/ui/test_organization.py::OrganizationTestCase::test_positive_update_name PASSED
tests/foreman/ui/test_organization.py::OrganizationTestCase::test_positive_update_ptable <- robottelo/decorators/__init__.py PASSED
tests/foreman/ui/test_organization.py::OrganizationTestCase::test_positive_update_smartproxy <- robottelo/decorators/__init__.py SKIPPED
tests/foreman/ui/test_organization.py::OrganizationTestCase::test_positive_update_template <- robottelo/decorators/__init__.py PASSED
tests/foreman/ui/test_organization.py::OrganizationTestCase::test_positive_update_user PASSED
tests/foreman/ui/test_organization.py::OrganizationTestCase::test_verify_bugzilla_1225588 <- robottelo/decorators/__init__.py FAILED
tests/foreman/ui/test_organization.py::OrganizationTestCase::test_verify_bugzilla_1259248 <- robottelo/decorators/__init__.py FAILED
```

From failed, only the ones related to location are related to this PR. Running them again:

```console
pytest tests/foreman/ui/test_organization.py -v -k "OrganizationTestCase and test_positive_create_with_location"
============================================================================= test session starts ==============================================================================
platform linux2 -- Python 2.7.12, pytest-3.0.7, py-1.4.32, pluggy-0.4.0 -- /home/renzo/PycharmProjects/venvs/robottelo/bin/python
cachedir: .cache
rootdir: /home/renzo/PycharmProjects/robottelo, inifile:
plugins: xdist-1.15.0, services-1.1.14, cov-2.4.0
collected 33 items 
2017-04-13 19:25:59 - conftest - DEBUG - Found WONTFIX in decorated tests ['1269196', '1378009', '1245334', '1221971', '1217635', '1226425', '1156555', '1199150', '1204686', '1267224', '1103157', '1230902', '1214312', '1079482']

2017-04-13 19:25:59 - conftest - DEBUG - Collected 33 test cases

2017-04-13 19:25:59 - conftest - DEBUG - Deselected test tests.foreman.ui.test_organization.test_positive_create_with_auto_populated_label due to WONTFIX


tests/foreman/ui/test_organization.py::OrganizationTestCase::test_positive_create_with_location <- robottelo/decorators/__init__.py PASSED

============================================================================= 32 tests deselected ==============================================================================
================================================================== 1 passed, 32 deselected in 113.80 seconds ===================================================================
```

```console
pytest tests/foreman/ui/test_organization.py -v -k "OrganizationTestCase and test_positive_update_location"
============================================================================= test session starts ==============================================================================
platform linux2 -- Python 2.7.12, pytest-3.0.7, py-1.4.32, pluggy-0.4.0 -- /home/renzo/PycharmProjects/venvs/robottelo/bin/python
cachedir: .cache
rootdir: /home/renzo/PycharmProjects/robottelo, inifile:
plugins: xdist-1.15.0, services-1.1.14, cov-2.4.0
collected 33 items 
2017-04-13 19:28:55 - conftest - DEBUG - Found WONTFIX in decorated tests ['1269196', '1378009', '1245334', '1221971', '1217635', '1226425', '1156555', '1199150', '1204686', '1267224', '1103157', '1230902', '1214312', '1079482']

2017-04-13 19:28:55 - conftest - DEBUG - Collected 33 test cases

2017-04-13 19:28:55 - conftest - DEBUG - Deselected test tests.foreman.ui.test_organization.test_positive_create_with_auto_populated_label due to WONTFIX


tests/foreman/ui/test_organization.py::OrganizationTestCase::test_positive_update_location <- robottelo/decorators/__init__.py PASSED

============================================================================= 32 tests deselected ==============================================================================
================================================================== 1 passed, 32 deselected in 245.85 seconds ===================================================================
```

abalakh commented:
So this time there's no mess with test steps - instead of creating magic helpers like `do_stuff()` which performed half of the test, you introduced helpers responsible for 1 logical test step - create organization with some entity, add entity to org, remove entity from org. Maybe this approach isn't that short (still reduces a lot of code duplication though), but you can see every single step of the test without the need to dig into code deeply.

oshtaier commented:
ACK
